consensus/src/chained_bft/block_storage/block_tree.rs

```diff
@@ -371,32 +372,29 @@ where
         });
 
         vote_msg.signature().clone().add_to_li(author, li_with_sig);
+        match validator_verifier.check_voting_power_and_signatures(li_with_sig.signatures().keys())
+        {
+            Ok(_) => {
+                let quorum_cert =
+                    QuorumCert::new(vote_msg.vote_data().clone(), li_with_sig.clone());
+                // Note that the block might not be present locally, in which case we cannot
+                // calculate time between block creation and qc
+                if let Some(time_to_qc) = self.get_block(&block_id).and_then(|block| {
+                    duration_since_epoch()
+                        .checked_sub(Duration::from_micros(block.timestamp_usecs()))
+                }) {
+                    counters::CREATION_TO_QC_S.observe_duration(time_to_qc);
+                }
 
-        let num_votes = li_with_sig.signatures().len();
-        if num_votes >= min_votes_for_qc {
-            let quorum_cert = QuorumCert::new(
-                VoteData::new(
-                    block_id,
-                    vote_msg.vote_data().executed_state_id(),
-                    vote_msg.vote_data().block_round(),
-                    vote_msg.vote_data().parent_block_id(),
-                    vote_msg.vote_data().parent_block_round(),
-                    vote_msg.vote_data().grandparent_block_id(),
-                    vote_msg.vote_data().grandparent_block_round(),
-                ),
-                li_with_sig.clone(),
-            );
-            // Note that the block might not be present locally, in which case we cannot calculate
-            // time between block creation and qc
-            if let Some(time_to_qc) = self.get_block(&block_id).and_then(|block| {
-                duration_since_epoch().checked_sub(Duration::from_micros(block.timestamp_usecs()))
-            }) {
-                counters::CREATION_TO_QC_S.observe_duration(time_to_qc);
+                VoteReceptionResult::NewQuorumCertificate(Arc::new(quorum_cert))
             }
-
-            return VoteReceptionResult::NewQuorumCertificate(Arc::new(quorum_cert));
+            Err(err) => match err {
+                VerifyError::TooLittleVotingPower { voting_power, .. } => {
+                    VoteReceptionResult::VoteAdded(voting_power)
+                }
+                _ => panic!("Illegal or unknown signature from vote_msg {}", vote_msg),
```

Comment:
we probably should not panic on a invalid vote, unless we claim we already verify it before hit this path